File: 2024/python/day01/part1.py
# part1.py
import sys


def main():
    """
    Main function to read the input file, process the entries, and calculate
    the total distance between location IDs.
    """
    if len(sys.argv) < 2:
        print("Usage: python part1.py <input_file>")
        sys.exit(1)
    filename = sys.argv[1]
    try:
        with open(filename) as file:
            total_distance = 0
            left_list = []
            right_list = []
            for line in file:
                entries = line.strip().split() # strips newlines and spilts by spaces
                if entries != []:
                    left_list.append(entries[0])
                    right_list.append(entries[1])
        file.close()

        left_list = sorted(left_list)
        right_list = sorted(right_list)

        # Iterate through the lists and calculate the total distance
        for i in range(len(left_list)):
            pair = (left_list[i], right_list[i])
            dist = abs(int(pair[0]) - int(pair[1]))
            total_distance += dist

        print(f"Total distance: {total_distance}")
    except FileNotFoundError:
        print(f"Error: The file {filename} was not found.")

if __name__ == "__main__":
    main()

File: 2024/cpp/day01/part1.cpp
#include <stdio.h>

#include <algorithm>  // sorting algo
#include <cmath>      // using abs()
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using std::cerr;
using std::cout;
using std::endl;
using std::ifstream;
using std::istringstream;
using std::string;
using std::vector;

int main(int argc, char* argv[]) {
    // Open input file for reading
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <input_file>" << endl;
        return 1;
    }
    ifstream filein(argv[1]);
    if (!filein) {
        cerr << "Error opening input file" << endl;
        return 1;
    }

    // initialize variables for left and right list, and line for file.
    string line;
    vector<int> left_list;
    vector<int> right_list;

    left_list.reserve(1000);
    right_list.reserve(1000);

    while (getline(filein, line)) {
        istringstream iss(line);  // stringstream treats like input source
        int left, right;

        if (iss >> left >> right) {
            left_list.push_back(left);
            right_list.push_back(right);
        }
    }

    // sort both lists
    std::sort(left_list.begin(), left_list.end());
    std::sort(right_list.begin(), right_list.end());

    // calculate pairs and distances
    int total_distance = 0;
    for (int i = 0; i < left_list.size(); i++) {
        total_distance += abs(left_list[i] - right_list[i]);
    }
    cout << "Total Distance: " << total_distance << endl;
    filein.close();  // close the file
    return 0;
}
